In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import pyodbc
import sqlalchemy
import sqlite3
from subprocess import check_output
import os

%sql sqlite://

'Connected: @None'

In [2]:
currencies = pd.read_csv('/kaggle/input/countriescurrencies/currencies.csv', 
                           index_col = 'code')
print(currencies.info())
currencies.head()

<class 'pandas.core.frame.DataFrame'>
Index: 224 entries, AFG to ZWE
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   curr_id        224 non-null    int64  
 1   basic_unit     224 non-null    object 
 2   curr_code      224 non-null    object 
 3   frac_unit      223 non-null    object 
 4   frac_perbasic  223 non-null    float64
dtypes: float64(1), int64(1), object(3)
memory usage: 10.5+ KB
None


curr_id      basic_unit curr_code frac_unit  frac_perbasic
code                                                            
AFG         1  Afghan afghani       AFN       Pul          100.0
ALB         2    Albanian lek       ALL  Qindarke          100.0
DZA         3  Algerian dinar       DZD   Santeem          100.0
AND         4            Euro       EUR      Cent          100.0
AGO         5  Angolan kwanza       AOA   Centimo          100.0

In [3]:
countries = pd.read_csv('/kaggle/input/countriescurrencies/countries.csv', index_col = 'code')
print(countries.info())
countries.head()

<class 'pandas.core.frame.DataFrame'>
Index: 206 entries, AFG to PSE
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   country_name  206 non-null    object 
 1   continent     206 non-null    object 
 2   region        206 non-null    object 
 3   surface_area  206 non-null    float64
 4   indep_year    188 non-null    float64
 5   local_name    206 non-null    object 
 6   gov_form      206 non-null    object 
 7   capital       201 non-null    object 
 8   cap_long      204 non-null    float64
 9   cap_lat       204 non-null    float64
dtypes: float64(4), object(6)
memory usage: 17.7+ KB
None


country_name continent                     region  surface_area  \
code                                                                      
AFG      Afghanistan      Asia  Southern and Central Asia      652090.0   
NLD      Netherlands    Europe             Western Europe       41526.0   
ALB          Albania    Europe            Southern Europe       28748.0   
DZA          Algeria    Africa            Northern Africa     2381740.0   
ASM   American Samoa   Oceania                  Polynesia         199.0   

      indep_year             local_name                 gov_form    capital  \
code                                                                          
AFG       1919.0  Afganistan/Afqanestan          Islamic Emirate      Kabul   
NLD       1581.0              Nederland  Constitutional Monarchy  Amsterdam   
ALB       1912.0              Shqiperia                 Republic     Tirane   
DZA       1962.0     Al-Jaza’ir/Algerie                 Republic    Algiers   
ASM          NaN          Amerika Samoa             US Territory  Pago Pago   

       cap_long  cap_lat  
code                      
AFG    69.17610  34.5228  
NLD     4.89095  52.3738  
ALB    19.81720  41.3317  
DZA     3.05097  36.7397  
ASM  -170.69100 -14.2846

In [4]:
cities = pd.read_csv('/kaggle/input/countriescurrencies/cities.csv', 
                          index_col = 'name')
print(cities.info())
cities.head()

<class 'pandas.core.frame.DataFrame'>
Index: 236 entries, Abidjan to Zunyi
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country_code     236 non-null    object 
 1   city_proper_pop  236 non-null    int64  
 2   metroarea_pop    137 non-null    float64
 3   urbanarea_pop    236 non-null    int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 9.2+ KB
None


country_code  city_proper_pop  metroarea_pop  urbanarea_pop
name                                                                   
Abidjan              CIV          4765000            NaN        4765000
Abu Dhabi            ARE          1145000            NaN        1145000
Abuja                NGA          1235880      6000000.0        1235880
Accra                GHA          2070463      4010054.0        2070463
Addis Ababa          ETH          3103673      4567857.0        3103673

In [5]:
economies = pd.read_csv('/kaggle/input/countriescurrencies/economies.csv', 
                           index_col = 'code')
print(economies.info())
economies.head()

<class 'pandas.core.frame.DataFrame'>
Index: 380 entries, AFG to ZWE
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   econ_id            380 non-null    int64  
 1   year               380 non-null    int64  
 2   income_group       380 non-null    object 
 3   gdp_percapita      378 non-null    float64
 4   gross_savings      338 non-null    float64
 5   inflation_rate     377 non-null    float64
 6   total_investment   344 non-null    float64
 7   unemployment_rate  217 non-null    float64
 8   exports            351 non-null    float64
 9   imports            349 non-null    float64
dtypes: float64(7), int64(2), object(1)
memory usage: 32.7+ KB
None


econ_id  year         income_group  gdp_percapita  gross_savings  \
code                                                                     
AFG         1  2010           Low income        539.667         37.133   
AFG         2  2015           Low income        615.091         21.466   
AGO         3  2010  Upper middle income       3599.270         23.534   
AGO         4  2015  Upper middle income       3876.200         -0.425   
ALB         5  2010  Upper middle income       4098.130         20.011   

      inflation_rate  total_investment  unemployment_rate  exports  imports  
code                                                                         
AFG            2.179            30.402                NaN   46.394   24.381  
AFG           -1.549            18.602                NaN  -49.110   -7.294  
AGO           14.480            14.433                NaN   -3.266  -21.076  
AGO           10.287             9.552                NaN    6.721  -21.778  
ALB            3.605            31.305               14.0   10.645   -8.013

In [6]:
languages = pd.read_csv('/kaggle/input/countriescurrencies/languages.csv', 
                           index_col = 'code')
print(languages.info())
languages.head()

<class 'pandas.core.frame.DataFrame'>
Index: 955 entries, AFG to ZWE
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   lang_id   955 non-null    int64  
 1   name      955 non-null    object 
 2   percent   518 non-null    float64
 3   official  955 non-null    bool   
dtypes: bool(1), float64(1), int64(1), object(1)
memory usage: 30.8+ KB
None


lang_id      name  percent  official
code                                      
AFG         1      Dari     50.0      True
AFG         2    Pashto     35.0      True
AFG         3    Turkic     11.0     False
AFG         4     Other      4.0     False
ALB         5  Albanian     98.8      True

In [7]:
populations = pd.read_csv('/kaggle/input/countriescurrencies/populations.csv', 
                           index_col = 'pop_id')
print(populations.info())
populations.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 434 entries, 20 to 432
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   country_code     434 non-null    object 
 1   year             434 non-null    int64  
 2   fertility_rate   402 non-null    float64
 3   life_expectancy  398 non-null    float64
 4   size             433 non-null    float64
dtypes: float64(3), int64(1), object(1)
memory usage: 20.3+ KB
None


country_code  year  fertility_rate  life_expectancy        size
pop_id                                                                
20              ABW  2010           1.704        74.953537    101597.0
19              ABW  2015           1.647        75.573585    103889.0
2               AFG  2010           5.746        58.970829  27962207.0
1               AFG  2015           4.653        60.717171  32526562.0
12              AGO  2010           6.416        50.654171  21219954.0

In [8]:
from sqlalchemy import create_engine
engine = create_engine('sqlite:////COUNTRIES_INFO_DETAILS', echo=False)
countries.to_sql('countries', con = engine)
economies.to_sql('economies', con = engine)
currencies.to_sql('currencies', con = engine)
cities.to_sql('cities', con = engine)
languages.to_sql('languages', con = engine)
populations.to_sql('populations', con = engine)

# PRACTICEs

## 1. LEFT JOIN

### Exercise 1.1.

Now you'll explore the differences between performing an inner join and a left join using the cities and countries tables.

You'll begin by performing an inner join with the cities table on the left and the countries table on the right. Remember to alias the name of the city field as city and the name of the country field as country.

You will then change the query to a left join. Take note of how many records are in each query here!

**Step 1.** Fill in the code based on the instructions in the code comments to complete the inner join. Note how many records are in the result of the join in the query result tab.

In [9]:
pd.read_sql(
    """
    SELECT c1.country_name AS country, code, c2.name AS city,
           region, city_proper_pop
           
    -- From left table (with alias)
    FROM countries AS c1
    
    -- Join to right table (with alias)
    INNER JOIN cities AS c2
    
    -- Match on country code
    ON c1.code = c2.country_code
    
    -- Order by descending country code
    ORDER BY code DESC;
    """, con = engine)

country code        city                     region  \
0                Zimbabwe  ZWE      Harare             Eastern Africa   
1                  Zambia  ZMB      Lusaka             Eastern Africa   
2            South Africa  ZAF   Cape Town            Southern Africa   
3            South Africa  ZAF      Durban            Southern Africa   
4            South Africa  ZAF  Ekurhuleni            Southern Africa   
..                    ...  ...         ...                        ...   
225             Argentina  ARG     Rosario              South America   
226  United Arab Emirates  ARE   Abu Dhabi                Middle East   
227  United Arab Emirates  ARE       Dubai                Middle East   
228                Angola  AGO      Luanda             Central Africa   
229           Afghanistan  AFG       Kabul  Southern and Central Asia   

     city_proper_pop  
0            1606000  
1            1742979  
2            3740026  
3            3442361  
4            3178470  
..               ...  
225          1193605  
226          1145000  
227          2643410  
228          2825311  
229          3414100  

[230 rows x 5 columns]

In the first step; there are 230 observations contained in the output

**Step 2.** Change the code to perform a **LEFT JOIN** instead of an **INNER JOIN**. 

After executing this query, note how many records the query result contains.

In [10]:
pd.read_sql(
    """ 
    SELECT c1.name AS city, code, c2.country_name AS country,
           region, city_proper_pop
    FROM cities AS c1
    -- 1. Join right table (with alias)
    LEFT JOIN countries AS c2
    -- 2. Match on country code
    ON c1.country_code = c2.code
    -- 3. Order by descending country code
    ORDER BY code DESC;
    """, con = engine)

city  code       country           region  city_proper_pop
0             Harare   ZWE      Zimbabwe   Eastern Africa          1606000
1             Lusaka   ZMB        Zambia   Eastern Africa          1742979
2          Cape Town   ZAF  South Africa  Southern Africa          3740026
3             Durban   ZAF  South Africa  Southern Africa          3442361
4         Ekurhuleni   ZAF  South Africa  Southern Africa          3178470
..               ...   ...           ...              ...              ...
231        Kaohsiung  None          None             None          2778918
232  New Taipei City  None          None             None          3954929
233         Taichung  None          None             None          2752413
234           Tainan  None          None             None          1885252
235           Taipei  None          None             None          2704974

[236 rows x 5 columns]

Hence, the **INNER JOIN** version resulted in `230 records`. The **LEFT JOIN** version returned `236 rows`.


### Exercise 1.2

Next, you'll try out another example comparing an inner join to its corresponding left join. Before you begin though, take note of how many records are in both the countries and languages tables below.

You will begin with an inner join on the countries table on the left with the languages table on the right. Then you'll change the code to a left join in the next bullet.

Note the use of multi-line comments here using `/* and */.`

Firstly, perform an `inner join`. Alias the name of the `country` field as `country` and the `name` of the `language` field as `language`.

Sort based on descending country name.

In [11]:
pd.read_sql(
    """
        /*
        Select country name AS country, the country's local name,
        the language name AS language, and
        the percent of the language spoken in the country
        */
        SELECT c.country_name AS country, local_name, l.name AS language, percent
        -- 1. From left table (alias as c)
        FROM countries AS c
        -- 2. Join to right table (alias as l)
        INNER JOIN languages AS l
        -- 3. Match on fields
        ON l.code = c.code
        -- 4. Order by descending country
        ORDER BY country DESC;
    """, con = engine)

country             local_name     language  percent
0       Zimbabwe               Zimbabwe        Shona      NaN
1       Zimbabwe               Zimbabwe      Ndebele      NaN
2       Zimbabwe               Zimbabwe      English      NaN
3       Zimbabwe               Zimbabwe        Chewa      NaN
4       Zimbabwe               Zimbabwe     Chibarwe      NaN
..           ...                    ...          ...      ...
909      Albania              Shqiperia  unspecified      0.1
910  Afghanistan  Afganistan/Afqanestan         Dari     50.0
911  Afghanistan  Afganistan/Afqanestan       Pashto     35.0
912  Afghanistan  Afganistan/Afqanestan       Turkic     11.0
913  Afghanistan  Afganistan/Afqanestan        Other      4.0

[914 rows x 4 columns]

Perform a left join instead of an inner join. Observe the result, and also note the change in the number of records in the result.

Carefully review which records appear in the left join result, but not in the inner join result.

In [12]:
pd.read_sql(
    """
        /*
        Select country name AS country, the country's local name,
        the language name AS language, and
        the percent of the language spoken in the country
        */
        SELECT c.country_name AS country, local_name, l.name AS language, percent
        -- 1. From left table (alias as c)
        FROM countries AS c
        -- 2. Join to right table (alias as l)
        LEFT JOIN languages AS l
        -- 3. Match on fields
        ON l.code = c.code
        -- 4. Order by descending country
        ORDER BY country DESC;
    """, con = engine)

country             local_name     language  percent
0       Zimbabwe               Zimbabwe        Shona      NaN
1       Zimbabwe               Zimbabwe      Ndebele      NaN
2       Zimbabwe               Zimbabwe      English      NaN
3       Zimbabwe               Zimbabwe        Chewa      NaN
4       Zimbabwe               Zimbabwe     Chibarwe      NaN
..           ...                    ...          ...      ...
916      Albania              Shqiperia  unspecified      0.1
917  Afghanistan  Afganistan/Afqanestan         Dari     50.0
918  Afghanistan  Afganistan/Afqanestan       Pashto     35.0
919  Afghanistan  Afganistan/Afqanestan       Turkic     11.0
920  Afghanistan  Afganistan/Afqanestan        Other      4.0

[921 rows x 4 columns]

So, in this exercise; the **INNER JOIN** version resulted in `914` records. The LEFT JOIN version returned `921` rows.

### Exercise 1.3.

You'll now revisit the use of the AVG() function introduced in our Intro to SQL for Data Science course. You will use it in combination with left join to determine the average gross domestic product (GDP) per capita by region in 2010.

**Step 1.** Begin with a left join with the countries table on the left and the economies table on the right.

Focus only on records with 2010 as the year.

In [13]:
pd.read_sql(
    """
        -- Select name, region, and gdp_percapita
        SELECT c.country_name, c.region, e.gdp_percapita
        -- 1. From countries (alias as c)
        FROM countries AS c
        -- 2. Left join with economies (alias as e)
        LEFT JOIN economies AS e
        -- 3. Match on code fields
        ON c.code = e.code
        -- 4. Focus on 2010
        WHERE e.year = 2010;
    """, con = engine)

country_name                     region  gdp_percapita
0             Afghanistan  Southern and Central Asia        539.667
1                  Angola             Central Africa       3599.270
2                 Albania            Southern Europe       4098.130
3    United Arab Emirates                Middle East      34628.630
4               Argentina              South America      10412.950
..                    ...                        ...            ...
180                 Samoa                  Polynesia       3434.100
181                 Yemen                Middle East       1266.790
182          South Africa            Southern Africa       7361.940
183                Zambia             Eastern Africa       1456.160
184              Zimbabwe             Eastern Africa        765.418

[185 rows x 3 columns]

**Step 2.**

In [14]:
pd.read_sql(
    """ 
        -- Select fields
        SELECT region, AVG(gdp_percapita) AS avg_gdp
        -- From countries (alias as c)
        FROM countries AS c
        -- Left join with economies (alias as e)
        LEFT JOIN economies AS e
        -- Match on code fields
        ON c.code = e.code
        -- Focus on 2010
        WHERE year = 2010
        -- Group by region
        GROUP BY region;
    """, con = engine)

region       avg_gdp
0   Australia and New Zealand  44792.385000
1            Baltic Countries  12631.030000
2             British Islands  43588.330000
3                   Caribbean  11413.339462
4              Central Africa   4797.239889
5             Central America   4969.970000
6              Eastern Africa   1757.348187
7                Eastern Asia  26205.851667
8              Eastern Europe  10095.456667
9                   Melanesia   2532.610000
10                 Micronesia   4497.652000
11                Middle East  18204.641765
12           Nordic Countries  57073.998000
13              North America  47911.510000
14            Northern Africa   4713.861667
15                  Polynesia   3451.363333
16              South America   7420.599167
17             Southeast Asia  10547.154400
18            Southern Africa   5051.598000
19            Southern Europe  22926.410909
20  Southern and Central Asia   2797.155000
21             Western Africa    998.954875
22             Western Europe  58130.962857

In [15]:
pd.read_sql(
    """
        -- Select fields
        SELECT region, AVG(gdp_percapita) AS avg_gdp
        -- From countries (alias as c)
        FROM countries AS c
        -- Left join with economies (alias as e)
        LEFT JOIN economies AS e
        -- Match on code fields
        ON c.code = e.code
        -- Focus on 2010
        WHERE year = 2010
        -- Group by region
        GROUP BY region
        -- Order by descending avg_gdp
        ORDER BY avg_gdp DESC;
    """, con = engine)

region       avg_gdp
0              Western Europe  58130.962857
1            Nordic Countries  57073.998000
2               North America  47911.510000
3   Australia and New Zealand  44792.385000
4             British Islands  43588.330000
5                Eastern Asia  26205.851667
6             Southern Europe  22926.410909
7                 Middle East  18204.641765
8            Baltic Countries  12631.030000
9                   Caribbean  11413.339462
10             Southeast Asia  10547.154400
11             Eastern Europe  10095.456667
12              South America   7420.599167
13            Southern Africa   5051.598000
14            Central America   4969.970000
15             Central Africa   4797.239889
16            Northern Africa   4713.861667
17                 Micronesia   4497.652000
18                  Polynesia   3451.363333
19  Southern and Central Asia   2797.155000
20                  Melanesia   2532.610000
21             Eastern Africa   1757.348187
22             Western Africa    998.954875

## 2. Right join
Right joins aren't as common as left joins. One reason why is that you can always write a right join as a `left join`.

#### Exercise. Convert this code to use RIGHT JOINs instead of LEFT JOINs 

                        SELECT cities.name AS city, urbanarea_pop, countries.name AS country,
                               indep_year, languages.name AS language, percent
                        FROM cities
                          LEFT JOIN countries
                            ON cities.country_code = countries.code
                          LEFT JOIN languages
                            ON countries.code = languages.code
                        ORDER BY city, language;

**Instruction**

The left join code is commented out here. Your task is to write a new query using rights joins that produces the same result as what the query using left joins produces. Keep this left joins code commented as you write your own query just below it using right joins to solve the problem.

Note the order of the joins matters in your conversion to using right joins!

        SELECT cities.name AS city, urbanarea_pop, countries.country_name AS country,
               indep_year, languages.name AS language, percent
        FROM languages
        RIGHT JOIN countries
        ON languages.code = countries.code
        RIGHT JOIN cities
        ON countries.code = cities.country_code
        ORDER BY city, language;


In [16]:
pd.read_sql(
    """
        SELECT cities.name AS city, urbanarea_pop, countries.country_name AS country,
               indep_year, languages.name AS language, percent
        FROM cities
        LEFT JOIN countries
        ON cities.country_code = countries.code
        LEFT JOIN languages
        ON countries.code = languages.code
        ORDER BY city, language;
    """, con = engine)

city  urbanarea_pop               country  indep_year language  \
0       Abidjan        4765000         Cote d'Ivoire      1960.0   French   
1       Abidjan        4765000         Cote d'Ivoire      1960.0    Other   
2     Abu Dhabi        1145000  United Arab Emirates      1971.0   Arabic   
3     Abu Dhabi        1145000  United Arab Emirates      1971.0  English   
4     Abu Dhabi        1145000  United Arab Emirates      1971.0    Hindi   
...         ...            ...                   ...         ...      ...   
1370      Zunyi        6127009                 China     -1523.0   Minnan   
1371      Zunyi        6127009                 China     -1523.0    Other   
1372      Zunyi        6127009                 China     -1523.0       Wu   
1373      Zunyi        6127009                 China     -1523.0    Xiang   
1374      Zunyi        6127009                 China     -1523.0      Yue   

      percent  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
...       ...  
1370      NaN  
1371      NaN  
1372      NaN  
1373      NaN  
1374      NaN  

[1375 rows x 6 columns]

## 3.Full join
In this exercise, you'll examine how your results differ when using a full join versus using a left join versus using an inner join with the countries and currencies tables.

You will focus on the `North American` region and also where the name of the country is missing. Dig in to see what we mean!

Begin with a full join with countries on the left and currencies on the right. The fields of interest have been SELECTed for you throughout this exercise.

Then complete a similar left join and conclude with an inner join.

**Step 1.**

In [17]:
pd.read_sql(
    """
    SELECT name AS country, code, region, basic_unit
    -- 3. From countries
    FROM countries
    -- 4. Join to currencies
    FULL JOIN currencies
    -- 5. Match on code
    USING (code)
    -- 1. Where region is North America or null
    WHERE region = 'North America' OR region IS NULL
    -- 2. Order by region
    ORDER BY region;
    """, con = engine)

country code         region              basic_unit
0         Bermuda  BMU  North America        Bermudian dollar
1   United States  USA  North America    United States dollar
2          Canada  CAN  North America         Canadian dollar
3       Greenland  GRL  North America                     NaN
4             NaN  TMP            NaN    United States dollar
5             NaN  FLK            NaN  Falkland Islands pound
6             NaN  HKG            NaN        Hong Kong dollar
7             NaN  AIA            NaN   East Caribbean dollar
8             NaN  NIU            NaN      New Zealand dollar
9             NaN  ROM            NaN            Romanian leu
10            NaN  SHN            NaN      Saint Helena pound
11            NaN  SGS            NaN           British pound
12            NaN  TWN            NaN       New Taiwan dollar
13            NaN  WLF            NaN               CFP franc
14            NaN  MSR            NaN   East Caribbean dollar
15            NaN  IOT            NaN    United States dollar
16            NaN  CCK            NaN       Australian dollar
17            NaN  COK            NaN      New Zealand dollar

**Step 2.** Repeat the same query as above but use a LEFT JOIN instead of a FULL JOIN. Note what has changed compared to the FULL JOIN result!

In [18]:
pd.read_sql(
    """
        SELECT country_name AS country, code, region, basic_unit
    -- 1. From countries
    FROM countries
      -- 2. Join to currencies
      LEFT JOIN currencies
        -- 3. Match on code
        USING (code)
    -- 4. Where region is North America or null
    WHERE region = 'North America' OR region IS NULL
    -- 5. Order by region
    ORDER BY region;
    """, con = engine)

country code         region            basic_unit
0        Bermuda  BMU  North America      Bermudian dollar
1      Greenland  GRL  North America                  None
2         Canada  CAN  North America       Canadian dollar
3  United States  USA  North America  United States dollar

**Step 3.** Repeat the same query as above but use an INNER JOIN instead of a FULL JOIN. Note what has changed compared to the FULL JOIN and LEFT JOIN results!

In [19]:
pd.read_sql(
        """        
        SELECT country_name AS country, code, region, basic_unit
        FROM countries
          -- 1. Join to currencies
          INNER JOIN currencies
            USING (code)
        -- 2. Where region is North America or null
        WHERE region = 'North America' OR region IS NULL
        -- 3. Order by region
        ORDER BY region;
        """, con = engine)

country code         region            basic_unit
0        Bermuda  BMU  North America      Bermudian dollar
1         Canada  CAN  North America       Canadian dollar
2  United States  USA  North America  United States dollar

##### Comments: The FULL JOIN query returned 17 rows, the OUTER JOIN returned 4 rows, and the INNER JOIN only returned 3 rows.

#### Exercise 3.2. 
You'll now investigate a similar exercise to the last one, but this time focused on using a table with more records on the left than the right. You'll work with the languages and countries tables.

Begin with a full join with languages on the left and countries on the right. Appropriate fields have been selected for you again here.

**Step 1.**
Choose records in which countries.name starts with the capital letter 'V' or is NULL.

Arrange by countries.name in ascending order to more clearly see the results.

In [20]:
pd.read_sql(
    """ 
            SELECT countries.name, code, languages.name AS language
            -- 3. From languages
            FROM languages
            -- 4. Join to countries
            FULL JOIN countries
            -- 5. Match on code
            USING (code)
            -- 1. Where countries.name starts with V or is null
            WHERE countries.name LIKE 'V%' OR countries.name IS NULL
            -- 2. Order by ascending countries.name
            ORDER BY countries.name;
    """, con = engine)

name code          language
0                   Vanuatu  VUT  Tribal Languages
1                   Vanuatu  VUT           English
2                   Vanuatu  VUT            French
3                   Vanuatu  VUT             Other
4                   Vanuatu  VUT           Bislama
5                 Venezuela  VEN           Spanish
6                 Venezuela  VEN        indigenous
7                   Vietnam  VNM        Vietnamese
8                   Vietnam  VNM           English
9                   Vietnam  VNM             Other
10  Virgin Islands, British  VGB               NaN
11     Virgin Islands, U.S.  VIR               NaN
12                      NaN  HKG     Other Chinese
13                      NaN  HKG             Other
14                      NaN  MSR           English
15                      NaN  NIU            Niuean
16                      NaN  NIU            Niuean
17                      NaN  NIU           English
18                      NaN  NIU            Niuean
19                      NaN  NIU             Other
20                      NaN  NFK           English
21                      NaN  NFK             Other
22                      NaN  ROM          Romanian
23                      NaN  ROM         Hungarian
24                      NaN  ROM            Romani
25                      NaN  ROM             Other
26                      NaN  ROM       unspecified
27                      NaN  SPM            French
28                      NaN  TWN          Mandarin
29                      NaN  TWN         Taiwanese
30                      NaN  TWN             Hakka
31                      NaN  TKL         Tokelauan
32                      NaN  TKL           English
33                      NaN  TKL            Samoan
34                      NaN  TKL          Tuvaluan
35                      NaN  TKL          Kiribati
36                      NaN  AIA           English
37                      NaN  TKL              none
38                      NaN  TKL       unspecified
39                      NaN  WLF         Wallisian
40                      NaN  WLF          Futunian
41                      NaN  WLF            French
42                      NaN  WLF             Other
43                      NaN  ESH          Standard
44                      NaN  ESH         Hassaniya
45                      NaN  ESH          Moroccan
46                      NaN  TKL             Other
47                      NaN  CXR           English
48                      NaN  CXR           Chinese
49                      NaN  CXR             Malay
50                      NaN  CCK             Malay
51                      NaN  CCK           English
52                      NaN  COK           English
53                      NaN  COK        Rarotongan
54                      NaN  COK             Other
55                      NaN  HKG         Cantonese
56                      NaN  HKG           English
57                      NaN  HKG          Mandarin

**Step 2.**

In [21]:
pd.read_sql(
    """
        SELECT countries.country_name, code, languages.name AS language
        FROM languages
        LEFT JOIN countries
        USING (code)
        WHERE countries.country_name LIKE 'V%' OR countries.country_name IS NULL
        ORDER BY countries.country_name;
    """, con = engine)

country_name code          language
0          None  AIA           English
1          None  CXR           English
2          None  CXR           Chinese
3          None  CXR             Malay
4          None  CCK             Malay
5          None  CCK           English
6          None  COK           English
7          None  COK        Rarotongan
8          None  COK             Other
9          None  MSR           English
10         None  NIU            Niuean
11         None  NIU            Niuean
12         None  NIU           English
13         None  NIU            Niuean
14         None  NIU             Other
15         None  NFK           English
16         None  NFK             Other
17         None  ROM          Romanian
18         None  ROM         Hungarian
19         None  ROM            Romani
20         None  ROM             Other
21         None  ROM       unspecified
22         None  SPM            French
23         None  TWN          Mandarin
24         None  TWN         Taiwanese
25         None  TWN             Hakka
26         None  TKL         Tokelauan
27         None  TKL           English
28         None  TKL            Samoan
29         None  TKL          Tuvaluan
30         None  TKL          Kiribati
31         None  TKL             Other
32         None  TKL              none
33         None  TKL       unspecified
34         None  WLF         Wallisian
35         None  WLF          Futunian
36         None  WLF            French
37         None  WLF             Other
38         None  ESH          Standard
39         None  ESH         Hassaniya
40         None  ESH          Moroccan
41      Vanuatu  VUT  Tribal Languages
42      Vanuatu  VUT           Bislama
43      Vanuatu  VUT           English
44      Vanuatu  VUT            French
45      Vanuatu  VUT             Other
46    Venezuela  VEN           Spanish
47    Venezuela  VEN        indigenous
48      Vietnam  VNM        Vietnamese
49      Vietnam  VNM           English
50      Vietnam  VNM             Other

**Step 3.**

In [22]:
pd.read_sql(
    """ 
        SELECT countries.country_name, code, languages.name AS language
        FROM languages
        INNER JOIN countries
        USING (code)
        WHERE countries.country_name LIKE 'V%' OR countries.country_name IS NULL
        ORDER BY countries.country_name;
    """, con = engine)

country_name code          language
0      Vanuatu  VUT  Tribal Languages
1      Vanuatu  VUT           Bislama
2      Vanuatu  VUT           English
3      Vanuatu  VUT            French
4      Vanuatu  VUT             Other
5    Venezuela  VEN           Spanish
6    Venezuela  VEN        indigenous
7      Vietnam  VNM        Vietnamese
8      Vietnam  VNM           English
9      Vietnam  VNM             Other

Again, make sure to compare the number of records the different types of joins return and try to verify whether the results make sense.

In [23]:
pd.read_sql(
    """
            SELECT c1.name AS country, region, l.name AS language,
                    c2.basic_unit, c2.frac_unit
            -- 1. From countries (alias as c1)
            FROM countries AS c1
            -- 2. Join with languages (alias as l)
            FULL JOIN languages AS l
            -- 3. Match on code
            USING (code)
            -- 4. Join with currencies (alias as c2)
            FULL JOIN currencies AS c2
            -- 5. Match on code
            USING (code)
            -- 6. Where region like Melanesia and Micronesia
            WHERE region LIKE 'M%esia';    
    """, con = engine)

country      region                language  \
0                          Kiribati  Micronesia                 English   
1                          Kiribati  Micronesia                Kiribati   
2                  Marshall Islands  Micronesia                   Other   
3                  Marshall Islands  Micronesia             Marshallese   
4                             Nauru  Micronesia                   Other   
5                             Nauru  Micronesia                 English   
6                             Nauru  Micronesia                 Nauruan   
7                     New Caledonia   Melanesia                   Other   
8                     New Caledonia   Melanesia                  French   
9                             Palau  Micronesia                   Other   
10                            Palau  Micronesia                 Chinese   
11                            Palau  Micronesia                Filipino   
12                            Palau  Micronesia                 English   
13                            Palau  Micronesia                   Other   
14                            Palau  Micronesia                 Palauan   
15                 Papua New Guinea   Melanesia                   Other   
16                 Papua New Guinea   Melanesia                    Hiri   
17                 Papua New Guinea   Melanesia                 English   
18                 Papua New Guinea   Melanesia               Tok Pisin   
19                  Solomon Islands   Melanesia              indigenous   
20                  Solomon Islands   Melanesia                 English   
21                  Solomon Islands   Melanesia       Melanesian pidgin   
22                          Vanuatu   Melanesia                   Other   
23                          Vanuatu   Melanesia                  French   
24                          Vanuatu   Melanesia                 English   
25                          Vanuatu   Melanesia                 Bislama   
26                          Vanuatu   Melanesia        Tribal Languages   
27  Micronesia, Federated States of  Micronesia          Kapingamarangi   
28  Micronesia, Federated States of  Micronesia                 Nukuoro   
29  Micronesia, Federated States of  Micronesia                Woleaian   
30  Micronesia, Federated States of  Micronesia                Ulithian   
31  Micronesia, Federated States of  Micronesia                  Yapese   
32  Micronesia, Federated States of  Micronesia               Pohnpeian   
33  Micronesia, Federated States of  Micronesia                 Kosrean   
34  Micronesia, Federated States of  Micronesia                Chuukese   
35  Micronesia, Federated States of  Micronesia                 English   
36                     Fiji Islands   Melanesia                     NaN   
37         Northern Mariana Islands  Micronesia                   Other   
38         Northern Mariana Islands  Micronesia             Other Asian   
39         Northern Mariana Islands  Micronesia                 Chinese   
40         Northern Mariana Islands  Micronesia    Other Pacific Island   
41         Northern Mariana Islands  Micronesia                 English   
42         Northern Mariana Islands  Micronesia                Chamorro   
43         Northern Mariana Islands  Micronesia              Philippine   
44                             Guam  Micronesia                   Other   
45                             Guam  Micronesia                   Asian   
46                             Guam  Micronesia  Other Pacific Islander   
47                             Guam  Micronesia                Chamorro   
48                             Guam  Micronesia                Filipino   
49                             Guam  Micronesia                 English   

                basic_unit frac_unit  
0        Australian dollar      Cent  
1        Australian dollar      Cent  
2     United States dollar      Cent  
3     United States dollar      Cent  
4      

## 4. Crossing

### 4.1 Table of 2 cities

This exercise looks to explore languages potentially and most frequently spoken in the cities of Hyderabad, India and Hyderabad, Pakistan.

You will begin with a cross join with cities AS c on the left and languages AS l on the right. Then you will modify the query using an inner join in the next tab.

**Step 1.** 
Create the cross join as described above. (Recall that cross joins do not use ON or USING.)

Make use of LIKE and Hyder% to choose Hyderabad in both countries.

Select only the city name AS city and language name AS language.

In [24]:
pd.read_sql(
    """ 
        SELECT c.name AS city, l.name AS language
        -- 1. From cities (alias as c)
        FROM cities AS c        
        -- 2. Join to languages (alias as l)
        CROSS JOIN languages AS l
        WHERE c.name LIKE 'Hyder%';
    """, con = engine)

city  language
0             Hyderabad      Dari
1             Hyderabad    Pashto
2             Hyderabad    Turkic
3             Hyderabad     Other
4             Hyderabad  Albanian
...                 ...       ...
1905  Hyderabad (India)     Sotho
1906  Hyderabad (India)     Tonga
1907  Hyderabad (India)    Tswana
1908  Hyderabad (India)     Venda
1909  Hyderabad (India)     Xhosa

[1910 rows x 2 columns]

**Step 2.**

In [25]:
pd.read_sql(
    """
        SELECT c.name AS city, l.name AS language
        FROM cities AS c        
        INNER JOIN languages AS l
        ON l.code = c.country_code
        WHERE c.name LIKE 'Hyder%';
    """, con = engine)

city    language
0   Hyderabad (India)       Hindi
1   Hyderabad (India)     Bengali
2   Hyderabad (India)      Telugu
3   Hyderabad (India)     Marathi
4   Hyderabad (India)       Tamil
5   Hyderabad (India)        Urdu
6   Hyderabad (India)    Gujarati
7   Hyderabad (India)     Kannada
8   Hyderabad (India)   Malayalam
9   Hyderabad (India)       Oriya
10  Hyderabad (India)     Punjabi
11  Hyderabad (India)    Assamese
12  Hyderabad (India)    Maithili
13  Hyderabad (India)       Other
14          Hyderabad     Punjabi
15          Hyderabad      Sindhi
16          Hyderabad     Saraiki
17          Hyderabad      Pashto
18          Hyderabad        Urdu
19          Hyderabad     Balochi
20          Hyderabad      Hindko
21          Hyderabad      Brahui
22          Hyderabad     English
23          Hyderabad  Burushaski
24          Hyderabad       Other

### Outer challenge

Now that you're fully equipped to use outer joins, try a challenge problem to test your knowledge!

In terms of life expectancy for 2010, determine the names of the lowest twenty countries and their regions.

In [26]:
pd.read_sql(
    """     
        SELECT c.country_name AS country, c.region, p.life_expectancy AS life_exp
        FROM countries AS c
        LEFT JOIN populations AS p
        ON c.code = p.country_code
        WHERE year = 2010
        ORDER BY life_exp
        LIMIT 20
    """, con = engine)

country           region   life_exp
0                    Andorra  Southern Europe        NaN
1             American Samoa        Polynesia        NaN
2             Cayman Islands        Caribbean        NaN
3                   Dominica        Caribbean        NaN
4                  Gibraltar  Southern Europe        NaN
5      Saint Kitts and Nevis        Caribbean        NaN
6                     Monaco   Western Europe        NaN
7           Marshall Islands       Micronesia        NaN
8   Northern Mariana Islands       Micronesia        NaN
9                      Nauru       Micronesia        NaN
10                     Palau       Micronesia        NaN
11  Turks and Caicos Islands        Caribbean        NaN
12                    Tuvalu        Polynesia        NaN
13   Virgin Islands, British        Caribbean        NaN
14                   Lesotho  Southern Africa  47.483415
15  Central African Republic   Central Africa  47.625317
16              Sierra Leone   Western Africa  48.228951
17                 Swaziland  Southern Africa  48.345756
18                  Zimbabwe   Eastern Africa  49.574659
19                      Chad   Central Africa  49.862073

In [ ]:
pd.read_sql(
    """     
        SELECT c.country_name AS country, c.region, p.life_expectancy AS life_exp
        FROM countries AS c
        LEFT JOIN populations AS p
        ON c.code = p.country_code
        WHERE year = 2010
        ORDER BY life_exp
        LIMIT 50
    """, con = engine)